### PR TITLE
caching: memcache optimizations

### PR DESCRIPTION
**Pull request by kujenga**
These optimizations are intended to help improve efficiency in a few specific situations. The local locking capability allows us to eliminate the memcache-based locking that normally doubles calls to memcache in order to eliminate duplicate retrievals across instances.

The concurrent fills allows for improved handling of transient errors while filling the cache, allowing the cache to send of multiple retrievals, any of which can fill the cache, allowing some to fail and still let operations continue.

### CODE REVIEW

**Comment by kujenga** [comment]
I've seen a few timeouts in running these tests, I'm still not fully convinced this works as intended... Putting a WIP label on it in the meantime.

**Comment by kujenga** [comment]
There was a bug in the `ConcurrentFills` functionality where unlock was non-deterministic, a given operation would use a separate randomly generated key for lock and unlock, which may or may not be the same. This is now fixed using an unlocking closure that captures the generated key.

Testing also revealed a race condition present in the Context cacher, which has been resolved with an additional `RWMutex` to protect the cache for a given `context.Context`.

This should be good to go!

File: lock/lock.go [hunk 1]
```diff
@@ -108,15 +109,30 @@ func WaitAndAcquire(ctx context.Context, key string, lockDuration, timeout time.
...
+			case <-done:
+				return
+			default:
+				// random sleep of 50-100ms
+				// TODO: make this configurable, add optional backoff functionality
+				time.Sleep((time.Duration)(50+rand.Int63n(50)) * time.Millisecond)
+				ticker <- struct{}{}
+			}
+		}
+	}()
```
Comment (AndyHaskell) [inline lock/lock.go]: What was the reasoning for rolling your own ticker here? If it's to have the ticker do a random wait time btw, maybe that can be its own type in Gometa!

File: caching/context.go [hunk 2]
```diff
@@ -103,14 +90,49 @@ func (c *Context) getItem() *ctxMapItem {
...
+			mp:      &map[interface{}]interface{}{},
+			RWMutex: &sync.RWMutex{},
+			mut:     &sync.Mutex{},
 		}, nil
 	})
 
 	return &elem
 }
 
+// helper type stpring all the values for a context
```
Comment (AndyHaskell) [inline caching/context.go]: Typo here

File: lock/lock.go (hunk 1 above)
Comment (kujenga) [inline lock/lock.go]: In the end it wasn't strictly necessary, when I implemented it I was suspicious there was some sort of deadlock going on with the locking, but that ended up not being the case.

Still, it seems like randomization might give slightly better properties to the behavior of this package, for example spreading out a batch of locks that all came in at very similar times.

I could make it it's own type, but in doing so I'd probably want to wrap it up in a struct and give a nice testable interface, perhaps that's better suited for a future PR?

Comment (AndyHaskell) [inline lock/lock.go]: Yeah a future PR was what I was thinking

File: caching/memcache.go [hunk 3]
```diff
@@ -15,7 +19,33 @@ import (
...
+	// another call must be made serially in order to fill the cache. By allowing
+	// multiple concurrent fills, any successfull response will fill the cache.
+	ConcurrentFills int
+	// LocalLock indicates whether or not the user wants locking on concurrent
+	// cache fills to occur only locally. The default behavior is to lock using
+	// memcache globally, but this doubles the number of memcache calls made.
+	LocalLock bool
+
+	mut    sync.Mutex // for serializing mutmap access
+	mutmap map[string]*retainMut
```
Comment (AndyHaskell) [inline caching/memcache.go]: So to get this straight, this `mutmap` is used specifically to have a mutex for each key when we're doing __local__ lock for a `Memcache`?

**Comment by AndyHaskell** [comment]
:+1:

**Pull request merged**